### DISCUSSION TITLE
vba-articles/issues-with-vba

### DISCUSSION THREAD

**Discussion by giscus** [sancarn/vba-articles#3] [_giscus]
# vba-articles/issues-with-vba

This post is still under development but here’s a dump of all issues. Hopefully I’ll get around to finishing it some day 👀 Issues with the language Object creation and method call is slow when compared to modern languages (and compared to module only code). Hidden features which are unimplementable - e.g. unable to implement IEnumVARIANT. I.E. No custom For each ... in ... next implementations (unless you delegate to a collection) Some low level standard interfaces are forbidden in VBA (like IDispatch) Inability to define hidden methods. Inability to use Evaluate methods on custom classes unless you initially cast to Object (IDispatch) which also removes complete intellisense… Lack of component based design for UserForms. - Modern UI frameworks are so much better at this by now! Lack of standard libraries: a. Lack of a canvas component for UserForms - Fundamentally limits what you can do in a userform. b. No standard built-in implementation of HashMap - Dictionary is great, but has to be included externally and isn’t available on Mac. Libraries can also help here. c. Lack of standard libraries - there are community solutions for this e.g. stdVBA and vb core lib Inconsistent setting of variables - the set keyword was added because Microsoft wanted to allow for use of default properties on COM objects. It would have been better if these were symbolised by illegal code e.g. something@() Inconsistent call conventions for subs and functions ByRef is default where ByVal is more logical. Assume this was initially to keep code optimal, but generally leads to difficulty in learning. People often just learn “Use ByVal everywhere”, which isn’t correct either. Inability to define collective types e.g. Collection<Car>. This leads to limitations in the type system. No built in lambda syntax. Understandable due to VBA’s age, but modern languages use a lot of Lambda syntax to make code cleaner. To get around this, TarVK and I reinvented the wheel and created our own lambda syntax systems to get around this fundamental flaw in the language. Inability to Multithread (or perform tasks asynchronously without re-writing the runtime) Unable to compute on the GPU natively. Lack of native libraries forces people to use APIs which only work on single platforms - Yes VBA can run on Mac but due to the lack of Windows APIs on Mac, what you can do with VBA is severely limited. Lack of true inheritance (can sort of fake it with defaults) No overloading - not a requirement, but a nice feature. Interfacing - Lack of implicit cast to interfaces can make use somewhat clumsy: Dim x as class: set x = new class: Dim y as IClass: set y = x: y.poop() Poor error reporting - Lack of native stack traces, lack of line numbers in errors Lack of reflection, metaprogramming and dynamic dependency injection. Structs/Types appear to have been bodged on top of the runtime. No union types exist and recursive types don’t exist either. PtrSafe keyword - Ridiculous that this only exists in some later versions of VBA… Lack of Events on base types - e.g. Collection::Add(), Collection::Remove() VBA Keywords - Print, Write, Debug, … - All of these are methods you cannot use! This wouldn’t be the case if VBA namespace wasn’t globally accessible. Why was it designed this way?! Inability to pass structs BYVAL to low level functions Introduction of the pointless keyword PtrSafe - This keyword provides no function whatsoever nor any guarantee of safety. Other issues with VBAs Environment: Not VBA’s fault, but the Excel/Word/Powerpoint object libraries are a mess… Error handling in VBA looks bad because generally error handling in Excel/Word/Powerpoint APIs is awful. Lack of Excel/Word/Powerpoint API events. Not VBA’s fault, but the Macro recorder produces garbage code. It’s really useful for testing, and it’s arguably very flexible, but that also means macros recorded with it (and based on current state) can really screw up other spreadsheets accidentally if something out of the ordinary happens. Not VBA’s fault, but the VBE - although it was great once - it’s now unmaintained and frankly awful. Rubberduck is good if you can install addins etc. but if you aren’t able to do these things then you’re stuck with the rubbish editor which is in there currently… Limited integration with new features such as PowerQuery, OfficeJS, etc. VB7 (VBA) never released officially to VB6 users. This means many libraries don’t work. It also means VB7 cannot exist as a standalone application. It must always be hosted by some other application. VBE doesn’t make it super easy to work outside of itself (lol). Limitations in Office (e.g. Excel limited rows etc.) Lack of type hinting Cannot easily run VBA in the cloud. Cannot easily schedule VBA scripts either. Though, see TwinBasic. Benefits of VBA VBA is on every computer with Microsoft Office. You don’t need permission to run it, everyone can use it VBA can access the local file system unlike OfficeJS / OfficeScript. VBA has Excel APIs which are mostly intuitive VBA has the ability to reference any other registered type library on the system. VBA can use native APIs on Windows OS, and thus can be used to automate windows. VBA is implemented on top of COM, which means VBA objects can be used by other languages, VBA can inject machine code into memory and execute it (can use thunks) Conditional compiling - Massive benefit to be able to “execute code” at compile time. Default object methods can be useful for adding syntax sugar to code. Ability to pass any datatype by reference can be very beneficial. Ability to create raw array datatypes. Enables full user interaction with WinAPI Implementation of LongPtr which can work between both 64-bit and 32 bit systems More reading: https://rubberduckvba.wordpress.com/2019/04/10/whats-wrong-with-vba/

https://sancarn.github.io/vba-articles/issues-with-vba.html

**Comment by munrobasher** [comment]
One thing I really miss is class constructors.
And the VB quick way of initialising objects, e.g. Dim Person As New Person("Rob")

**Reply by sancarn** [reply]
Thanks, you're right, that should be added! Not sure how they were missed!

**Reply by munrobasher** [reply]
I suspect because they'd decided not to develop VBA any further. Which is bizarre because they developed VB.NET which IMO isn't that bad if you want to stick with BASIC core syntax.
